pkg/device/kunlunxin/device.go

```diff
@@ -0,0 +1,262 @@
+/*
+Copyright 2024 The HAMi Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kunlunxin
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/Project-HAMi/HAMi/pkg/util"
+	"github.com/Project-HAMi/HAMi/pkg/util/nodelock"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+)
+
+const (
+	XPUDevice      = "XPU"
+	XPUCommonWord  = "XPU"
+	NodeLock       = "hami.io/mutex.lock"
+	RegisterAnnos  = "hami.io/node-register-xpu"
+	HandshakeAnnos = "hami.io/node-handshake-xpu"
+	UseUUIDAnno    = "hami.io/use-xpu-uuid"
+	NoUseUUIDAnno  = "hami.io/no-use-xpu-uuid"
+)
+
+var (
+	XpuResourceCount  string
+	XpuResourceMemory string
+)
+
+type Devices struct {
+}
+
+func ParseConfig(fs *flag.FlagSet) {
+	fs.StringVar(&XpuResourceCount, "xpu-name", "kunlunxin.com/xpu", "kunlunxin resource name")
+	fs.StringVar(&XpuResourceMemory, "xpu-memory-name", "kunlunxin.com/xpu-memory", "kunlunxin resource memory name")
+}
+
+type XPUConfig struct {
+	XpuResourceName       string `yaml:"resourceCountName"`
+	XpuResourceMemoryName string `yaml:"resourceMemoryName"`
+}
+
+func InitXPUDevice(config XPUConfig) *Devices {
+	XpuResourceCount = config.XpuResourceName
+	XpuResourceMemory = config.XpuResourceMemoryName
+	util.InRequestDevices[XPUDevice] = "hami.io/xpu-devices-to-allocate"
+	util.SupportDevices[XPUDevice] = "hami.io/xpu-devices-allocated"
+	util.HandshakeAnnos[XPUDevice] = HandshakeAnnos
+	return &Devices{}
+}
+
+func (dev *Devices) trimMemory(m int64) int64 {
+	temps := []int64{24576, 49152}
```

Comment:
Is this because xpu can only achieve three slicing specifications: 1, 1/2 and 1/4?